composer/loggers/neptune_logger.py

```diff
@@ -0,0 +1,295 @@
+
+import os
+import pathlib
+import warnings
+from functools import partial
+from typing import TYPE_CHECKING, Any, Dict, Optional, Sequence, Union
+
+import numpy as np
+import torch
+
+from composer._version import __version__
+from composer.loggers import LoggerDestination
+from composer.utils import MissingConditionalImportError, dist
+
+if TYPE_CHECKING:
+    from composer import Logger
+    from composer.core import State
+
+
+class NeptuneLogger(LoggerDestination):
+    """LoggerDestination implementation that sends the logs to Neptune.
+
+    Args:
+        project (str, optional): The name of your Neptune project, in the form `workspace-name/project-name`.
+            If you leave it empty, the NEPTUNE_PROJECT environment variable will be used.
+        api_token (str, optional): Your Neptune API token. You can leave out this argument if you save your token to the
+            NEPTUNE_API_TOKEN environment variable (recommended).
+            You can find your API token in the user menu of the Neptune web app.
+        rank_zero_only (bool, optional): Whether to log only on the rank-zero process. Default: ``True``.
+        upload_artifacts (bool, optional): Whether the logger should upload artifacts to Neptune. Default: ``False``.
+        base_namespace (str, optional): The name of the base namespace to log the metadata to. Default: "training".
+        neptune_kwargs (Dict[str, Any], optional): Any additional keyword arguments to the neptune.init_run()
+            function. For options, see the [Run API reference](https://docs.neptune.ai/api/neptune/#init_run) in the
+            Neptune docs.
+
+    Example:
+        Default Neptune logger:
+            from composer import Trainer
+            from composer.loggers import NeptuneLogger
+            neptune_logger = NeptuneLogger()
+            trainer = Trainer(loggers=neptune_logger)
+            trainer.fit()
+
+        More options:
+            from composer import Trainer
+            from composer.loggers import NeptuneLogger
+            neptune_logger = NeptuneLogger(
+                project="ml-team/classification",
+                upload_artifacts=True,
+                dependencies="infer")  # Neptune Run kwarg
+            trainer = Trainer(loggers=neptune_logger)
+            trainer.fit()
+
+    For more, see the [Neptune-Composer integration guide](https://docs.neptune.ai/integrations/composer/).
+    """
+    METRIC_NAMESPACE = 'metrics'
+    HYPERPARAM_NAMESPACE = 'hyperparameters'
+    TRACE_NAMESPACE = 'traces'
+    INTEGRATION_VERSION_KEY = 'source_code/integrations/neptune-MosaicML'
+
+    def __init__(
+        self,
+        *,
+        project: Optional[str] = None,
+        api_token: Optional[str] = None,
+        rank_zero_only: bool = True,
+        upload_artifacts: bool = False,
+        base_namespace: str = 'training',
+        **neptune_kwargs,
+    ) -> None:
+        try:
+            from neptune.internal.utils import verify_type
```

Comment:
It's not going to work with `neptune-client<1.0.0`, right? Shouldn't we add support for it?